src/shogun/kernel/DotKernel.h

```diff
@@ -61,6 +61,7 @@ class CDotKernel : public CKernel
 		virtual bool init(CFeatures* l, CFeatures* r)
 		{
 			CKernel::init(l,r);
+			initialise_auto_params();
```

Comment:
rather than calling it here, why not call it just after the lambda was attached to the parameter. Then it would be like "normal" parameters being initialised, usually in the private "init" method where they all receive defaults and are registered ...

Reply:
sorry I have to take this back, obviously what I wrote doesn't work as the features are not available.

Reply:
so devs need to be careful and put this everywhere where the member that the value depends on would be changed ... for kernel that is clearly "init" so all good. Maybe we should make it impossible to change "lhs" and "rhs" via put in that case.... thoughts?

Reply:
I think actually a parameter property that is “read-only” and then a generic check in put might be a good idea. Then we can “force” users to use the unit method if they wanna use a kernel (and the auto parameters are safe). This is a pattern we had used before but with constructors and the lack of setters. Thoughts?

Reply:
> so devs need to be careful and put this everywhere where the member that the value depends on would be changed ...
Yes, that is annoying but there is no other way I can think of

> I think actually a parameter property that is “read-only” and then a generic check in put might be a good idea.

Hmm I am not sure I understand what you mean. The parameter property is "read-only" no? And I added a check in put that verifies that a parameter has been initialised. Can you write a short pseudo code?

> This is a pattern we had used before but with constructors and the lack of setters.

Can you link it and I will have a look as I am not sure I understand what you mean?